### PR TITLE
Updated and included init.d config

### PR DESCRIPTION
**Pull request by BClark09**
Added init.d to the gradle build for Linux init systems based on sysVinit.

Also prevented more than one instance from running "start" multiple times and implemented a simple mechanism to keep /var/run/openhab.pid up to date. 

This _should not_ affect Linux systems based on systemd.

Resolves https://github.com/openhab/openhab-distro/issues/328

Signed-off-by: Ben Clark <ben@benjyc.uk>

### CODE REVIEW

**Comment by BClark09** [comment]
@theoweiss @ThomDietrich, could you comment on the efficiency of the proposed script if you have time?

File: resources/deb/etc/init.d/openhab2 [hunk 1]
```diff
@@ -71,15 +71,36 @@ fi
...
         log_daemon_msg "Starting $DESC" "$NAME"
-        #${OPENHAB_INST_DIR}/bin/setpermissions.sh > /dev/null
-        if start-stop-daemon --start --quiet --pidfile $PIDFILE -c "${USER_AND_GROUP}" --umask 002 --background --make-pidfile --chdir ${OPENHAB_DIR} \
-                --oknodo --exec /usr/share/openhab2/runtime/bin/start
+
+        # If a process still exists but there's no PID file, then openHAB is probably still shutting down. We should proceed.
+        if [ -z "$EXISTINGPID" ] || [ ! -f $PIDFILE ];
         then
-            log_end_msg 0
+            if start-stop-daemon --start --quiet --pidfile $PIDFILE -c "${USER_AND_GROUP}" --umask 002 --background --chdir ${OPENHAB_DIR} \
```
Comment (ThomDietrich) [inline resources/deb/etc/init.d/openhab2]: barely cosmetic: as there is already a second line, maybe you want to introduce another split here? It's longer than the GitHub preview, might be a bad sign =)

File: resources/deb/etc/init.d/openhab2 [hunk 2]
```diff
@@ -71,15 +71,36 @@ fi
 # Load the VERBOSE setting and other rcS variables
 . /lib/init/vars.sh
 
+# Check to see if openHAB is currently running
+EXISTINGPID=`pgrep java -G openhab | tail -1`
```
Comment (ThomDietrich) [inline resources/deb/etc/init.d/openhab2]: Good idea. I'm unsure if the filter is precise enough though. Yes the chances are pretty low that another java process is executed as openhab but still... What do you think?

File: resources/deb/etc/init.d/openhab2 [hunk 3]
```diff
@@ -71,15 +71,36 @@ fi
...
+                done
+                echo "$pid" > $PIDFILE
+                log_end_msg 0
+            else
+                log_end_msg 1
+            fi
         else
+            echo -n "Error: OpenHAB is already running with PID: $EXISTINGPID!"
+            # Update PID File incase it is stale.
+            echo "$EXISTINGPID" > $PIDFILE
```
Comment (ThomDietrich) [inline resources/deb/etc/init.d/openhab2]: It's probably meaningful but I don't get the comment. Why is this needed? Is it a really a good idea to create the PIDFILE in case of this "error"? The way I understand it, this may cover up an unexpected state...

File: resources/deb/etc/init.d/openhab2 (hunk 2 above)
Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: If it is then the variable will contain the latest process that the _openhab_ group has opened. I think this is fine because processes aren't killed based on the contents on the file and it is merely an indicator of that openHAB is doing something.

File: resources/deb/etc/init.d/openhab2 (hunk 1 above)
Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: Ah, good idea. My only change of this line was to remove the --make-pidfile option. Since the line got shorter, it hadn't occured to me that the line was already too long to begin with!

File: resources/deb/etc/init.d/openhab2 (hunk 3 above)
Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: EXISTINGPID is updated each time init.d/openhab is run, but the only time that it can change between points is if openHAB restarts itself or if the user runs stop.sh/start.sh manually. This line is to make sure that the PIDFILE is updated and valid whenever possible, as other processes that may monitor it need the up-to-date version.

File: resources/deb/etc/init.d/openhab2 (hunk 2 above)
Comment (theoweiss) [inline resources/deb/etc/init.d/openhab2]: I think we can not rely on pgrep being available on every debian/ubuntu installation from debian6, raspbian to current ubuntu, can we? A good hint is to check if other system services use this tool. If not, we shouldn't use it and have to stick with the basic unix/linux tools and the tools from init-functions.

Comment (ThomDietrich) [inline resources/deb/etc/init.d/openhab2]: good point! an important topic. maybe this helps http://unix.stackexchange.com/questions/234709/avoid-pgrep-in-sh

File: resources/deb/etc/init.d/openhab2 (hunk 3 above)
Comment (theoweiss) [inline resources/deb/etc/init.d/openhab2]: IMHO the init script shouldn't try to fix an inconsistent system state. It should be kept as small and simple as possible and focus on the task of (trying) to start/stop the service and get the status of the service. The script may assume that the service is only handled by the init script itself.
If an admin has started openHAB without using the init script he should be able to fix it on his own.

Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: Fair enough! Thanks both, I will remove the line and it's comment!

**Comment by BClark09** [comment]
Thanks for the comments @theoweiss  and @ThomDietrich! I updated to use `ps aux --sort=start_time | grep openhab.*java | grep -v grep | awk '{print $2}' | tail -1` so that any successful start should be update the pid file with the latest openhab owned java pid.

**Comment by ThomDietrich** [comment]
I've tested the line on my system, looks good. I'm still unsure about the effect `--sort=start_time` and `tail -1` will have. Can there be multiple processes, what should be done in this case, how can we be certain that the latest of these is the one we are looking for?

Also, how about only defining the command once?

**Comment by BClark09** [comment]
Thanks once again @ThomDietrich! 

 > Can there be multiple processes, what should be done in this case, how can we be certain that the latest of these is the one we are looking for?

The only time I was able to get multiple processes with this script was if you follow a stop command with a start command. In this case, the combination of `--sort=start_time` and `tail -1` will pick the starting process, as the first process will soon become invalid. 

Let's assume that there is some way that the openHAB user can open a second java process for itself: 

- `service openhab2 start` will function as expected, by preventing you from opening yet another instance of openHAB. The error message wouldn't technically be wrong either, any PID would still point to something that needs to be stopped before you should try and start another instance of openHAB.

- `service openhab2 stop` doesn't care about the PID number so it'll just call the stop script which should handle the correct stop procedure.

> Also, how about only defining the command once?

Good idea, this is best done as a function call right?

**Comment by ThomDietrich** [comment]
Okay then, I see you you thought of everything.
Function call or you could define the command as a string variable and call it via `eval` but maybe it's not worth it. I'd be emotionally okay if you just merged the script as it is now :smiley_cat:

**Comment by BClark09** [comment]
Always good to test the scripts (and my own) logic! @theoweiss if you are happy with this implementation and it's style/formatting then I would consider this safe to merge. I've tested on Ubuntu 14.04 and Debian 8 (just in case), which use the two different startup processes we should support.

**Comment by theoweiss** [comment]
Uups, I added some comments on monday but forgot to finish and submit the review. Now the comments are lost. I'll  try it again this evening :-((

File: resources/deb/etc/init.d/openhab2 [hunk 4]
```diff
@@ -104,7 +125,7 @@ case "$1" in
             then
                 log_end_msg 0
             else
-               log_end_msg 1
+                log_end_msg 1
```
Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: The extra space was to bring the line to the same alignment used above.  4 spaces?

File: resources/deb/etc/init.d/openhab2 [hunk 5]
```diff
@@ -71,15 +71,36 @@ fi
...
+        # If a process still exists but there's no PID file, then openHAB is probably still shutting down. We should proceed.
+        if [ -z "$EXISTINGPID" ] || [ ! -f $PIDFILE ];
         then
-            log_end_msg 0
+            if start-stop-daemon --start --quiet --pidfile $PIDFILE -c "${USER_AND_GROUP}" --umask 002 --background --chdir ${OPENHAB_DIR} \
+                    --oknodo --exec /usr/share/openhab2/runtime/bin/start
+            then
+                timeout=0
+                # The correct process is only available once the wrapper scripts have finished, we should prevent the user from running again
+                until [ $timeout -eq 5 ] || [ ! -z "$pid"  ];
```
Comment (BClark09) [inline resources/deb/etc/init.d/openhab2]: I added a timeout of 5 seconds which should be long enough for the wrapper scripts to finish and to start the main Java process. Unfortuately make-pidfile seems to point to a temporary process and I found many resources that suggest making your own in this case.

Furthermore, the systemd version doesn't even make a pid file,  should we not worry about it here either? The service doesn't use it in stopping itself and were only adding it here for compatibility with other programs that may use pids.

**Comment by theoweiss** [comment]
LGTM so far. I'll merge this PR.
